proxy/envoy/config.go

```diff
@@ -116,54 +116,94 @@ func Generate(context *ProxyContext) *Config {
 	}
 }
 
-// build combines the outbound and inbound routes prioritizing the latter
+// buildListeners produces a list of listeners and referenced clusters
 // build also returns all inbound clusters, and outbound clusters referenced by TCP proxy
 // (due to lack of RDS support for TCP proxy filter, all referenced clusters in the routes
-// must be present, the policy application is still performed by CDS rather than local agent)
-func build(context *ProxyContext) ([]*Listener, Clusters) {
-	httpRouteConfigs, tcpRouteConfigs := buildRoutes(context)
+// must be present)
+func buildListeners(context *ProxyContext) (Listeners, Clusters) {
+	// query the services model
+	instances := context.Discovery.HostInstances(context.Addrs)
+	services := context.Discovery.Services()
 
 	// canonicalize listeners and collect inbound clusters
-	// all outbound clusters are served with CDS
-	clusters := make(Clusters, 0)
-	listeners := make([]*Listener, 0)
+	inbound, inClusters := buildInboundListeners(instances, context.MeshConfig)
+	outbound, outClusters := buildOutboundListeners(instances, services, context)
 
-	for port, routeConfig := range httpRouteConfigs {
-		hosts := routeConfig.VirtualHosts
-		sort.Slice(hosts, func(i, j int) bool { return hosts[i].Name < hosts[j].Name })
-		clusters = append(clusters, routeConfig.filterClusters(func(cluster *Cluster) bool {
-			return !cluster.outbound
-		})...)
-
-		filters := buildFaultFilters(routeConfig)
-
-		filters = append(filters, HTTPFilter{
-			Type:   "decoder",
-			Name:   "router",
-			Config: FilterRouterConfig{},
-		})
-
-		listener := &Listener{
-			Address: fmt.Sprintf("tcp://0.0.0.0:%d", port),
-			Filters: []*NetworkFilter{{
-				Type: "read",
-				Name: HTTPConnectionManager,
-				Config: &HTTPFilterConfig{
-					CodecType:  "auto",
-					StatPrefix: "http",
-					AccessLog: []AccessLog{{
-						Path: DefaultAccessLog,
-					}},
-					RDS: &RDS{
-						Cluster:         RDSName,
-						RouteConfigName: fmt.Sprintf("%d", port),
-						RefreshDelayMs:  1000,
-					},
-					Filters: filters,
-				},
-			}},
+	listeners := append(inbound, outbound...)
+	sort.Slice(listeners, func(i, j int) bool { return listeners[i].Address < listeners[j].Address })
+	clusters := append(inClusters, outClusters...).Normalize()
+
+	return listeners, clusters
+}
+
+// buildHTTPListener constructs a listener for the network interface address and port
+// Use "0.0.0.0" IP address to listen on all interfaces
+// RDS parameter controls whether to use RDS for the route updates.
+func buildHTTPListener(routeConfig *HTTPRouteConfig, ip string, port int, rds bool) *Listener {
```

Comment:
Inbound http routes are not using RDS. They are just simple routes to localhost.